vivisect/tools/graphutil.py

```diff
@@ -700,31 +701,36 @@ def stop(self):
         self.__go__ = False
 
     def watchdog(self, time):
-        # FIXME: make this use one thread, not N
         '''
         set a watchdog timer for path generation (if it takes too long to get another path)
         '''
-        self.wdt = threading.Thread(target=self.__wd, args=[time])
-        self.wdt.setDaemon = True
-        self.wdt.start()
-
-    def __wd(self, maxsec):
-        # FIXME: make this use one thread, not N
-        maxsec *=10
-        count = 0
-        while self.__go__:
-            time.sleep(.1)
-            self.__steplock.acquire()
+        self._wd_maxsec = time
+        self._wd_count = 0
+        if self.wdt == None:
+            self.wdt = threading.Thread(target=self.__wd)
+            self.wdt.setDaemon = True
+            self.wdt.start()
+
+    def __wd(self):
+        while True:
             try:
-                if not self.__update:
-                    count += 1
-                    if count > maxsec:
-                        self.stop()
-                        break
-            finally:
-                self.__steplock.release()
-
-            self.__update = False
+                while self.__go__:
+                    time.sleep(.1)
+                    self.__steplock.acquire()
```

Comment:
this really should be a "with block" ( insert old-dog-new-tricks joke here ):
```
with self.__steplock:
    dostuff()
```
no need for finally / release ( and no chance of missing it )